# Limpeza de dados

### Importando os dados

In [91]:
import pandas as pd
import matplotlib.pyplot as plt

def ler_tabela(nome):
    df = pd.read_parquet(f"../data/processed/{nome}.parquet") 
    print(nome, df.shape)
    return df


pessoas = ler_tabela("pessoas")
sinistros = ler_tabela("sinistros")
veiculos = ler_tabela("veiculos")

pessoas (23211, 31)
sinistros (789122, 50)
veiculos (1015991, 12)


## DF Pessoas

### Verificando tabelas

In [92]:
# analisando tabela
print(pessoas.info())

<class 'pandas.DataFrame'>
RangeIndex: 23211 entries, 0 to 23210
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id_sinistro               23211 non-null  int64  
 1   id_pessoa                 23211 non-null  int64  
 2   id_veiculo                17390 non-null  float64
 3   cod_ibge                  23211 non-null  int64  
 4   municipio                 23211 non-null  str    
 5   regiao_administrativa     23211 non-null  str    
 6   tipo_via                  23211 non-null  str    
 7   tipo_veiculo_vitima       17901 non-null  str    
 8   tipo_de_vitima            23211 non-null  str    
 9   modo_transporte_vitima    23211 non-null  str    
 10  sexo                      23211 non-null  str    
 11  idade                     22153 non-null  float64
 12  gravidade_lesao           23211 non-null  str    
 13  faixa_etaria_demografica  23211 non-null  str    
 14  faixa_etaria_lega

In [93]:
# identificando valores nulos
print(pessoas.isna().mean())
print(pessoas.isna().sum())

id_sinistro                 0.000000
id_pessoa                   0.000000
id_veiculo                  0.250786
cod_ibge                    0.000000
municipio                   0.000000
regiao_administrativa       0.000000
tipo_via                    0.000000
tipo_veiculo_vitima         0.228771
tipo_de_vitima              0.000000
modo_transporte_vitima      0.000000
sexo                        0.000000
idade                       0.045582
gravidade_lesao             0.000000
faixa_etaria_demografica    0.000000
faixa_etaria_legal          0.000000
profissao                   0.636336
grau_de_instrucao           1.000000
nacionalidade               1.000000
data_sinistro               0.000000
ano_sinistro                0.000000
mes_sinistro                0.000000
dia_sinistro                0.000000
ano_mes_sinistro            0.000000
data_obito                  0.000000
ano_obito                   0.000000
mes_obito                   0.000000
dia_obito                   0.000000
a

### Limpando os dados

In [94]:
# graus de instrucao e nacionalidade sem valores, inuteis para analise
print(pessoas.notna().sum())

pessoas = pessoas.drop(columns=["grau_de_instrucao", "nacionalidade"])

print(pessoas.info())

id_sinistro                 23211
id_pessoa                   23211
id_veiculo                  17390
cod_ibge                    23211
municipio                   23211
regiao_administrativa       23211
tipo_via                    23211
tipo_veiculo_vitima         17901
tipo_de_vitima              23211
modo_transporte_vitima      23211
sexo                        23211
idade                       22153
gravidade_lesao             23211
faixa_etaria_demografica    23211
faixa_etaria_legal          23211
profissao                    8441
grau_de_instrucao               0
nacionalidade                   0
data_sinistro               23211
ano_sinistro                23211
mes_sinistro                23211
dia_sinistro                23211
ano_mes_sinistro            23211
data_obito                  23211
ano_obito                   23211
mes_obito                   23211
dia_obito                   23211
ano_mes_obito               23211
local_obito                 23211
local_via     

In [95]:
# trasformando coluna em datetime
pessoas["data_sinistro"] = pd.to_datetime(pessoas["data_sinistro"], format="%d/%m/%Y")
pessoas["data_obito"] = pd.to_datetime(pessoas["data_obito"], format="%d/%m/%Y")

# ajustando data, diminuindo colunas de 8 para 2

pessoas = pessoas.drop(
    columns=["ano_sinistro", "mes_sinistro", "dia_sinistro",
        "ano_mes_sinistro", "ano_obito", "mes_obito", "dia_obito", "ano_mes_obito"
])

print(pessoas.columns)

Index(['id_sinistro', 'id_pessoa', 'id_veiculo', 'cod_ibge', 'municipio',
       'regiao_administrativa', 'tipo_via', 'tipo_veiculo_vitima',
       'tipo_de_vitima', 'modo_transporte_vitima', 'sexo', 'idade',
       'gravidade_lesao', 'faixa_etaria_demografica', 'faixa_etaria_legal',
       'profissao', 'data_sinistro', 'data_obito', 'local_obito', 'local_via',
       'tempo_sinistro_obito'],
      dtype='str')


### Analisando tabelas

In [96]:
pessoas.describe()

,id_sinistro,id_pessoa,id_veiculo,cod_ibge,idade,data_sinistro,data_obito,tempo_sinistro_obito
count,2.321100e+04,2.321100e+04,1.739000e+04,2.321100e+04,22153.000000,23211,23211,23211.000000
mean,2.519887e+06,4.947594e+04,1.227387e+06,3.532503e+06,42.138672,2024-01-26 22:48:28.103916,2024-01-28 12:31:43.993796,1.571712
min,2.456227e+06,5.000000e+00,3.300000e+01,3.500105e+06,0.000000,2022-01-01 00:00:00,2022-01-01 00:00:00,0.000000
25%,2.473315e+06,1.946850e+04,4.010025e+04,3.518404e+06,27.000000,2023-01-30 00:00:00,2023-01-31 00:00:00,0.000000
50%,2.491011e+06,4.035900e+04,8.384050e+04,3.534401e+06,40.000000,2024-02-18 00:00:00,2024-02-20 00:00:00,0.000000
75%,2.507902e+06,5.985850e+04,2.308217e+06,3.550308e+06,55.000000,2025-01-19 00:00:00,2025-01-22 00:00:00,0.000000
max,2.845098e+06,2.241871e+06,3.291987e+06,3.557303e+06,102.000000,2025-12-31 00:00:00,2025-12-31 00:00:00,30.000000
std,7.246166e+04,1.363951e+05,1.280152e+06,1.721030e+04,18.467158,NaN,NaN,4.422074


In [97]:
print(pessoas.shape)

(23211, 21)


In [98]:
print(pessoas.info())

<class 'pandas.DataFrame'>
RangeIndex: 23211 entries, 0 to 23210
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   id_sinistro               23211 non-null  int64         
 1   id_pessoa                 23211 non-null  int64         
 2   id_veiculo                17390 non-null  float64       
 3   cod_ibge                  23211 non-null  int64         
 4   municipio                 23211 non-null  str           
 5   regiao_administrativa     23211 non-null  str           
 6   tipo_via                  23211 non-null  str           
 7   tipo_veiculo_vitima       17901 non-null  str           
 8   tipo_de_vitima            23211 non-null  str           
 9   modo_transporte_vitima    23211 non-null  str           
 10  sexo                      23211 non-null  str           
 11  idade                     22153 non-null  float64       
 12  gravidade_lesao           232

In [99]:
analise_numeros = pessoas.select_dtypes(include="number").agg(["min", "max", "mean", "median"])
print(analise_numeros)

         id_sinistro     id_pessoa    id_veiculo      cod_ibge       idade  \
min     2.456227e+06  5.000000e+00  3.300000e+01  3.500105e+06    0.000000   
max     2.845098e+06  2.241871e+06  3.291987e+06  3.557303e+06  102.000000   
mean    2.519887e+06  4.947594e+04  1.227387e+06  3.532503e+06   42.138672   
median  2.491011e+06  4.035900e+04  8.384050e+04  3.534401e+06   40.000000   

        tempo_sinistro_obito  
min                 0.000000  
max                30.000000  
mean                1.571712  
median              0.000000  


### Descobertas
##### Nulos
tabelas com dados nulos
- id_veiculo = 0.25 | 5821
- tipo_veiculo_vitima = 0.22 | 5310 
- idade = 0.04 | 1058
- profissao = 0.63 | 14770
- grau_de_instrução = 1.00 | 23211
- nacionalidade = 1.00 | 23211

##### medias
- mortes mensais sinistro/obito de onde vem os 0.11 de diferença?
- mediana de idade de obitos - 40
- ano com mais mortes - 2024
- mes com mais mortes - agosto
- dia com mais mortes - 16
- tempo sinistro com 1.57 de media e 0 mediana, acredito que seja por falta de cadastro

In [100]:
# checkando dados em 2026
filtro_2026 = pessoas[(pessoas["data_sinistro"].dt.year == 2026) & (pessoas["data_obito"].dt.year == 2026)]

print(filtro_2026)

Empty DataFrame
Columns: [id_sinistro, id_pessoa, id_veiculo, cod_ibge, municipio, regiao_administrativa, tipo_via, tipo_veiculo_vitima, tipo_de_vitima, modo_transporte_vitima, sexo, idade, gravidade_lesao, faixa_etaria_demografica, faixa_etaria_legal, profissao, data_sinistro, data_obito, local_obito, local_via, tempo_sinistro_obito]
Index: []

[0 rows x 21 columns]


In [101]:
# investigando diferença de %0.11 na media de mortes pós sinistro 
filtro_tempo = pessoas["data_obito"] != pessoas["data_sinistro"]

print(filtro_tempo.value_counts().sort_index())

pessoas["diferenca_dias_obito"] = (pessoas["data_obito"] - pessoas["data_sinistro"]).dt.days
print(pessoas["diferenca_dias_obito"].value_counts().sort_index())
print(pessoas["diferenca_dias_obito"].agg(["mean", "median"]))

False    17733
True      5478
Name: count, dtype: int64
diferenca_dias_obito
0     17733
1      1640
2       506
3       464
4       368
5       278
6       258
7       210
8       212
9       160
10      145
11      132
12      150
13      105
14       96
15       80
16       74
17       55
18       42
19       66
20       52
21       52
22       50
23       40
24       36
25       44
26       31
27       37
28       33
29       24
30       38
Name: count, dtype: int64
mean      1.571712
median    0.000000
Name: diferenca_dias_obito, dtype: float64


In [102]:
# verificando congruência para idade maxima

pd.set_option({"display.width": 100, "display.max_columns":None})
analise_b = pessoas.loc[pessoas["idade"] == pessoas["idade"].max()]

print(analise_b)

       id_sinistro  id_pessoa  id_veiculo  cod_ibge  municipio       regiao_administrativa  \
14468      2482301      47014     43565.0   3518800  GUARULHOS  METROPOLITANA DE SÃO PAULO   

           tipo_via tipo_veiculo_vitima tipo_de_vitima modo_transporte_vitima      sexo  idade  \
14468  VIAS URBANAS           AUTOMOVEL     PASSAGEIRO              AUTOMOVEL  FEMININO  102.0   

      gravidade_lesao faixa_etaria_demografica faixa_etaria_legal profissao data_sinistro  \
14468           FATAL                   90 e +         80 ou mais       NaN    2024-07-28   

      data_obito               local_obito local_via  tempo_sinistro_obito  diferenca_dias_obito  
14468 2024-08-07  ESTABELECIMENTO DE SAUDE   PUBLICO                  10.0                    10  


## DF Sinistros

### Verificando dados

In [103]:
print(sinistros.shape)

(789122, 50)


In [104]:
print(sinistros.info())

<class 'pandas.DataFrame'>
RangeIndex: 789122 entries, 0 to 789121
Data columns (total 50 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   id_sinistro                         789122 non-null  int64  
 1   tipo_registro                       789122 non-null  str    
 2   data_sinistro                       789122 non-null  str    
 3   ano_sinistro                        789122 non-null  int64  
 4   mes_sinistro                        789122 non-null  int64  
 5   dia_sinistro                        789122 non-null  int64  
 6   hora_sinistro                       788256 non-null  str    
 7   ano_mes_sinistro                    789122 non-null  str    
 8   dia_da_semana                       789122 non-null  str    
 9   turno                               789122 non-null  str    
 10  logradouro                          787366 non-null  str    
 11  numero_logradouro                   7

In [105]:
print(sinistros.describe())

        id_sinistro   ano_sinistro   mes_sinistro   dia_sinistro       latitude      longitude  \
count  7.891220e+05  789122.000000  789122.000000  789122.000000  700809.000000  700803.000000   
mean   2.102028e+06    2023.490444       6.637768      15.683912     -22.864441     -47.403123   
std    4.432743e+05       1.104123       3.383723       8.762257       1.003697       1.375932   
min    1.256247e+06    2022.000000       1.000000       1.000000     -87.500000     -89.000000   
25%    1.709722e+06    2023.000000       4.000000       8.000000     -23.570787     -47.824724   
50%    2.165776e+06    2023.000000       7.000000      16.000000     -23.301937     -46.849941   
75%    2.503930e+06    2024.000000      10.000000      23.000000     -22.324196     -46.565820   
max    2.845098e+06    2025.000000      12.000000      31.000000     -19.805309     -23.518641   

           cod_ibge  qtd_pedestre  qtd_bicicleta  qtd_motocicleta  qtd_automovel    qtd_onibus  \
count  7.891220e+05

In [106]:
print(sinistros.head())

   id_sinistro       tipo_registro data_sinistro  ano_sinistro  mes_sinistro  dia_sinistro  \
0      1825253         NOTIFICACAO    01/01/2022          2022             1             1   
1      2309521  SINISTRO NAO FATAL    01/01/2022          2022             1             1   
2      1785362  SINISTRO NAO FATAL    01/01/2022          2022             1             1   
3      1282118         NOTIFICACAO    01/01/2022          2022             1             1   
4      1869705  SINISTRO NAO FATAL    01/01/2022          2022             1             1   

  hora_sinistro ano_mes_sinistro dia_da_semana      turno                        logradouro  \
0         17:47          2022/01        Sábado      TARDE                             RUA 5   
1         19:41          2022/01        Sábado      NOITE             AVENIDA MAJOR NICÁCIO   
2         00:50          2022/01        Sábado  MADRUGADA  AVENIDA DOUTOR ULISSES GUIMARÃES   
3         00:33          2022/01        Sábado  MADRUGA

In [107]:
print(sinistros.isna().describe())

       id_sinistro tipo_registro data_sinistro ano_sinistro mes_sinistro dia_sinistro  \
count       789122        789122        789122       789122       789122       789122   
unique           1             1             1            1            1            1   
top          False         False         False        False        False        False   
freq        789122        789122        789122       789122       789122       789122   

       hora_sinistro ano_mes_sinistro dia_da_semana   turno logradouro numero_logradouro tipo_via  \
count         789122           789122        789122  789122     789122            789122   789122   
unique             2                1             1       1          2                 2        1   
top            False            False         False   False      False             False    False   
freq          788256           789122        789122  789122     787366            785975   789122   

       tipo_local latitude longitude cod_ibge mu

In [108]:
print(sinistros.isna().sum())
print(sinistros.isna().mean())

id_sinistro                                0
tipo_registro                              0
data_sinistro                              0
ano_sinistro                               0
mes_sinistro                               0
dia_sinistro                               0
hora_sinistro                            866
ano_mes_sinistro                           0
dia_da_semana                              0
turno                                      0
logradouro                              1756
numero_logradouro                       3147
tipo_via                                   0
tipo_local                             16033
latitude                               88313
longitude                              88319
cod_ibge                                   0
municipio                                  0
regiao_administrativa                      0
administracao                           1958
conservacao                             1884
circunscricao                           1979
tp_sinistr

### Limpando dados

In [109]:
# Ajustando tabelas data hora
sinistros["data_hora"] = pd.to_datetime(sinistros["data_sinistro"] + '' + sinistros["hora_sinistro"], format="%d/%m/%Y%H:%M%S")
print(sinistros["data_hora"])

# excluindo tabelas soltas
sinistros = sinistros.drop(columns=["tp_sinistro_colisao_traseira", "ano_sinistro", "mes_sinistro", 
                                    "dia_sinistro", "data_sinistro", "hora_sinistro", "ano_mes_sinistro"])

0        2022-01-01 17:04:07
1        2022-01-01 19:04:01
2        2022-01-01 00:05:00
3        2022-01-01 00:03:03
4        2022-01-01 21:01:05
                 ...        
789117   2025-12-31 16:05:07
789118   2025-12-31 17:05:00
789119   2025-12-31 19:01:04
789120   2025-12-31 20:02:05
789121   2025-12-31 23:05:05
Name: data_hora, Length: 789122, dtype: datetime64[us]


## DF Veiculos

### Verificando dados

In [ ]:
print(veiculos.info())

### Limpando dados

In [ ]:
veiculos["data_sinistro"] = pd.to_datetime(veiculos["data_sinistro"], format="%d/%m/%Y")

veiculos = veiculos.drop(columns=["ano_sinistro", "mes_sinistro", "dia_sinistro", "ano_mes_sinistro"]) 
print(veiculos.info())

## Atualizando dados em parquet

In [118]:
tabela = {
    "pessoas": pessoas,
    "sinistros": sinistros,
    "veiculos": veiculos
}

for nome, df in tabela.items():
    df.to_parquet(f"../data/processed/{nome}.parquet")
    print(f"tabela: {nome}, shape: {df}.shape")

tabela: pessoas, shape:        id_sinistro  id_pessoa  id_veiculo  cod_ibge              municipio  \
0          2479624      44067   2298533.0   3516408        FRANCO DA ROCHA   
1          2459328       5429   2895304.0   3516200                 FRANCA   
2          2498579      66870     54289.0   3501905                 AMPARO   
3          2477534      44418     34873.0   3550308              SAO PAULO   
4          2464234      16811   2897159.0   3550308              SAO PAULO   
...            ...        ...         ...       ...                    ...   
23206      2711462      47300   2978823.0   3550308              SAO PAULO   
23207      2711700      81358   2979190.0   3543907              RIO CLARO   
23208      2711512      40941         NaN   3523107        ITAQUAQUECETUBA   
23209      2711439      36922         NaN   3524402                JACAREI   
23210      2711483      35364   2978868.0   3548708  SAO BERNARDO DO CAMPO   

            regiao_administrativa      